

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Project outline &mdash; BEELINE  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> BEELINE
          

          
            
            <img src="_static/beeline.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="BEELINE.html">BEELINE</a></li>
<li class="toctree-l1"><a class="reference internal" href="BoolODE.html">BoolODE</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">BEELINE</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Project outline</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/beeline-tutorial.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <p>This tutorial will first explain the structure of the BEELINE repository,
with a walkthrough of the different components that the user can customize.</p>
<div class="section" id="project-outline">
<h1>Project outline<a class="headerlink" href="#project-outline" title="Permalink to this headline">¶</a></h1>
<p>The BEELINE repository is structured as follows:</p>
<div class="code text highlight-default notranslate"><div class="highlight"><pre><span></span>BEELINE
|-- inputs/
|   `-- examples/
|       `-- GSD/
|           |--refNetwork.csv
|           |--PseudoTime.csv
|           `--ExpressionData.csv
|-- config-files/
|   `-- config.yaml
|-- BLRun/
|   |-- sinceritiesRunner.py
|   `-- ...
|-- BLPlot/
|   |-- NetworkMotifs.py
|   `-- CuratedOverview.py
|-- BLEval/
|   |-- parseTime.py
|    `-- ...
|-- Algorithms/
|     |-- SINCERITIES/
|     `-- ...
|-- LICENSE
|-- initialize.sh
|-- BLRunner.py
|-- BLEvaluator.py
|-- README.md
`-- requirements.txt
</pre></div>
</div>
</div>
<div class="section" id="input-datasets">
<span id="id1"></span><h1>Input Datasets<a class="headerlink" href="#input-datasets" title="Permalink to this headline">¶</a></h1>
<p>The sample input data set provided is generated by <a class="reference internal" href="BoolODE.html#boolode"><span class="std std-ref">BoolODE</span></a>
using the Boolean model of <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/26573569">Gonadal Sex Determination</a> as input.  Note that
this dataset has been pre-processed to produce three files that are
required in the BEELINE pipline.</p>
<ol class="arabic simple">
<li><cite>ExpressionData.csv</cite> contains the RNAseq data, with genes as
rows and cell IDs as columns. This file is a required input to the
pipline. Here is a <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/inputs/example/GSD/ExpressionData.csv">sample ExpressionData.csv file</a></li>
<li><cite>PseudoTime.csv</cite> contains the pseudotime values for the cells in
<cite>ExpressionData.csv</cite>.  We recommend using the Slingshot method to
obtain the pseudotime for a dataset. Many algorithms in the
pipeline require a pseudotime file as input. Here is a <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/inputs/example/GSD/PseudoTime.csv">sample
PseudoTime file</a>.</li>
<li><cite>refNetwork.csv</cite> contains the ground truth network underlying the
interactions between genes in <cite>ExpressionData.csv</cite>. Typically this
network is not available, and will have to be curated from various
Transcription Factor databases. While this file is not a
requirement to run the base pipeline, a reference network is
required to run some of the performance evaluations in
<a class="reference internal" href="BLEval.html#bleval"><span class="std std-ref">BLEval package</span></a>.  Here is a <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/inputs/example/GSD/refNetwork.csv">sample refNetwork.csv file</a></li>
</ol>
<p>The figure below shows the t-SNE visualization of the expression
data from the example dataset.</p>
<img alt="_images/SlingshotOutputVis.png" src="_images/SlingshotOutputVis.png" />
<p>This dataset shows a bifurcating
trajectory, as is evidenced by the part (a) of the figure, where
each ‘cell’ is colored by the timepoint at which it was sampled
in the simulation (the darker colors indicate earlier time points).
Clustering the simulation confirms the two trajectories, indicated
in red and blue respectively in part (b). Finally, running Slingshot
on this dataset and specifying the cluster of cells corresponding to
the early time points yields two pseudotime trajectories, shown in part (c).
For details on the generation of this simulated dataset, see <a class="reference internal" href="BoolODE.html#boolode"><span class="std std-ref">BoolODE</span></a>.</p>
<div class="admonition attention">
<p class="first admonition-title">Attention</p>
<p class="last">Please ensure that any input dataset you create is
comma separated, and contains the correct style of
column names.</p>
</div>
</div>
<div class="section" id="config-files">
<span id="configfiles"></span><h1>Config files<a class="headerlink" href="#config-files" title="Permalink to this headline">¶</a></h1>
<p>Beeline uses <a class="reference external" href="https://yaml.org/">YAML</a> files to allow users to
flexibly specify inputs and algorithm run parameters. A sample config
file is provided in <a class="reference external" href="https://github.com/Murali-group/Beeline/blob/master/config-files/config.yaml">here</a>. A
config file should contain at minimum</p>
<div class="code text highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">input_settings</span><span class="p">:</span>
    <span class="n">input_dir</span> <span class="p">:</span> <span class="s2">&quot;Base input directory name, relative to current working directory (recommended: inputs)&quot;</span>
    <span class="n">dataset_dir</span><span class="p">:</span> <span class="s2">&quot;Subdirectory of input_dir where the datasets are placed in&quot;</span>
    <span class="n">datasets</span><span class="p">:</span>
        <span class="o">-</span> <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;Dataset name&quot;</span>
          <span class="n">exprData</span><span class="p">:</span> <span class="s2">&quot;Expression Data filename&quot;</span>
          <span class="n">cellData</span><span class="p">:</span> <span class="s2">&quot;PseudoTime filename&quot;</span>
          <span class="n">trueEdges</span><span class="p">:</span> <span class="s2">&quot;Ground truth network filename&quot;</span>
    <span class="n">algorithms</span><span class="p">:</span>
        <span class="o">-</span> <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;Algorithm name&quot;</span>
          <span class="n">params</span><span class="p">:</span>
              <span class="c1"># Any other parameters that can be passed to</span>
              <span class="c1"># this algorithm</span>
              <span class="n">should_run</span><span class="p">:</span> <span class="p">[</span><span class="kc">True</span><span class="p">]</span> <span class="c1"># or False</span>

<span class="n">output_settings</span><span class="p">:</span>
    <span class="n">output_dir</span> <span class="p">:</span> <span class="s2">&quot;Base output directory name, relative to current working directory (recommended: outputs)&quot;</span>
    <span class="n">output_prefix</span><span class="p">:</span> <span class="s2">&quot;Prefix for writing evaluation summary files&quot;</span>
</pre></div>
</div>
<p>Apart from indicating the path to the base directory and the specific
folder containing the input, the config file indicates which
algorithms should be run, along with the parameters to be passed to
the algorithms, if any. For a list of parameters that the pipeline
currently passes to the algorithms, see  <a class="reference internal" href="BEELINE.html#algorithms"><span class="std std-ref">Details of supported algorithms</span></a>.  Finally,
the YAML file also specifies the output_dir where the outputs are written.
Note that the output directory structure under output_dir is same
as the input directory strucutre under input_dir. For example, if the config file
contains the following:</p>
<div class="code text highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">input_settings</span><span class="p">:</span>
    <span class="n">input_dir</span> <span class="p">:</span> <span class="s2">&quot;inputs&quot;</span>
    <span class="n">dataset_dir</span><span class="p">:</span> <span class="s2">&quot;example&quot;</span>
    <span class="n">datasets</span><span class="p">:</span>
        <span class="o">-</span> <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;GSD&quot;</span>
          <span class="n">exprData</span><span class="p">:</span> <span class="s2">&quot;ExpressionData.csv&quot;</span>
          <span class="n">cellData</span><span class="p">:</span> <span class="s2">&quot;PseudoTime.csv&quot;</span>
          <span class="n">trueEdges</span><span class="p">:</span> <span class="s2">&quot;refNetwork.csv&quot;</span>
    <span class="n">algorithms</span><span class="p">:</span>
        <span class="o">-</span> <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;PIDC&quot;</span>
          <span class="n">params</span><span class="p">:</span>
              <span class="n">should_run</span><span class="p">:</span> <span class="p">[</span><span class="kc">True</span><span class="p">]</span>
        <span class="o">-</span> <span class="n">name</span><span class="p">:</span> <span class="s2">&quot;SCODE&quot;</span>
          <span class="n">params</span><span class="p">:</span>
              <span class="n">should_run</span><span class="p">:</span> <span class="p">[</span><span class="kc">False</span><span class="p">]</span>
              <span class="n">z</span><span class="p">:</span> <span class="p">[</span><span class="mi">10</span><span class="p">]</span>
              <span class="n">nIter</span><span class="p">:</span> <span class="p">[</span><span class="mi">1000</span><span class="p">]</span>
              <span class="n">nRep</span><span class="p">:</span> <span class="p">[</span><span class="mi">6</span><span class="p">]</span>

<span class="n">output_settings</span><span class="p">:</span>
    <span class="n">output_dir</span> <span class="p">:</span> <span class="s2">&quot;outputs&quot;</span>
    <span class="n">output_prefix</span><span class="p">:</span> <span class="s2">&quot;GSD&quot;</span>
</pre></div>
</div>
<p>BEELINE would interepret this as follows:</p>
<ul class="simple">
<li>The input expression data file is located at <code class="docutils literal notranslate"><span class="pre">inputs/example/GSD/ExpressionData.csv</span></code>, the pseudotime file is located at <code class="docutils literal notranslate"><span class="pre">inputs/example/GSD/PseudoTime.csv</span></code>, and the reference network or true edges file is located at <code class="docutils literal notranslate"><span class="pre">inputs/example/GSD/refNetwork.csv</span></code>. Note that the paths are relative to the current working directory.</li>
<li>The algorithm specific inputs will be placed under <code class="docutils literal notranslate"><span class="pre">inputs/example/GSD/&lt;algorithm_name&gt;</span></code>. For example, for PIDC, the inputs will be placed under <code class="docutils literal notranslate"><span class="pre">inputs/example/GSD/PIDC/</span></code>. The SCODE algorithm will be skipped becuase the should_run flag is set to False.</li>
<li>The output folder structure will be similar to that of the inputs under output_dir. For example, the outputs obtained after running PIDC on this dataset will be placed under <code class="docutils literal notranslate"><span class="pre">outputs/example/GSD/PIDC/</span></code>.</li>
</ul>
<div class="admonition attention">
<p class="first admonition-title">Attention</p>
<p class="last">Please ensure that the YAML file is correctly indented!</p>
</div>
</div>
<div class="section" id="running-the-pipeline">
<h1>Running the pipeline<a class="headerlink" href="#running-the-pipeline" title="Permalink to this headline">¶</a></h1>
<p>Once the input dataset has been generated and formatted as described
in Section <a class="reference internal" href="#input-datasets"><span class="std std-ref">Input Datasets</span></a> , and the config file has been
created as described in <a class="reference internal" href="#configfiles"><span class="std std-ref">Config files</span></a>, the pipeline can be
executed by simply calling <code class="docutils literal notranslate"><span class="pre">BLRun.py</span></code> with the config file
specifying the inputs and the algorithms to run, passed using the
<code class="docutils literal notranslate"><span class="pre">--config</span></code> option which takes the path to the config file.</p>
<p>To run the pipeline, simply invoke</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">BLRunner</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">config</span> <span class="n">PATH</span><span class="o">/</span><span class="n">TO</span><span class="o">/</span><span class="n">CONFIG</span><span class="o">/</span><span class="n">FILE</span>
</pre></div>
</div>
<p>For details about the implementation of <a class="reference internal" href="BLRun.html#module-BLRun" title="BLRun"><code class="xref py py-class docutils literal notranslate"><span class="pre">BLRun</span></code></a> , see <span class="xref std std-ref">blrunguide</span> .</p>
</div>
<div class="section" id="running-the-evaluation-scripts">
<h1>Running the evaluation scripts<a class="headerlink" href="#running-the-evaluation-scripts" title="Permalink to this headline">¶</a></h1>
<p>Each algorithm outputs an inferred network in the form of a ranked edge list.
BEELINE implements a consistent interface using the config file in order to retrieve
the predictions of multiple algorithms and evaluate them using a variety of methods.</p>
<p>The evaluation of the inferred networks is done by calling the
<code class="docutils literal notranslate"><span class="pre">BLEvaluator.py</span></code> script.  Like the <code class="docutils literal notranslate"><span class="pre">BLRunner.py</span></code> script, the
Evaluator script takes the config file as input. Every  subsequent
option passed to this script calls a different evaluation script. For instance,
in order to analyze the AUROC and AUPRC values, and also analyze network motifs,
use the following command</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">BLEvaluator</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">config</span> <span class="n">PATH</span><span class="o">/</span><span class="n">TO</span><span class="o">/</span><span class="n">CONFIG</span><span class="o">/</span><span class="n">FILE</span> \
                           <span class="o">--</span><span class="n">auc</span> \ <span class="c1"># calls the computeAUC script</span>
                           <span class="o">--</span><span class="n">motifs</span> \ <span class="c1"># calls the computeNetMotifs script</span>
</pre></div>
</div>
<p>The full list of available evaluation functions and their corresponding options to
be passed to <code class="docutils literal notranslate"><span class="pre">BLEvaluator.py</span></code> are given below:</p>
<table border="1" class="colwidths-given docutils">
<colgroup>
<col width="38%" />
<col width="63%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>-h, –help</td>
<td>show the help message and exit</td>
</tr>
<tr class="row-even"><td>-c, –config &lt;file-name&gt;</td>
<td>Configuration file containing list of datasets, algorithms, and output specifications.</td>
</tr>
<tr class="row-odd"><td>-a, –auc</td>
<td>Compute median of areas under Precision-Recall and ROC curves. Calls <a class="reference internal" href="BLEval.html#module-BLEval.computeAUC" title="BLEval.computeAUC"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.computeAUC</span></code></a>.</td>
</tr>
<tr class="row-even"><td>-j, –jaccard</td>
<td>Compute median Jaccard index of predicted top-k networks
for each algorithm for a given set of datasets generated from the same
ground truth network. Calls <a class="reference internal" href="BLEval.html#module-BLEval.computeJaccard" title="BLEval.computeJaccard"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.computeJaccard</span></code></a>.</td>
</tr>
<tr class="row-odd"><td>-r, –spearman</td>
<td>Compute median Spearman Corr. of predicted edges for each algorithm for a given set of datasets generated from the same ground truth network.  Calls <a class="reference internal" href="BLEval.html#module-BLEval.computeSpearman" title="BLEval.computeSpearman"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.computeSpearman</span></code></a>.</td>
</tr>
<tr class="row-even"><td>-t, –time</td>
<td>Analyze time taken by each algorithm for a. Calls <a class="reference internal" href="BLEval.html#module-BLEval.parseTime" title="BLEval.parseTime"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.parseTime</span></code></a>.</td>
</tr>
<tr class="row-odd"><td>-e, –epr</td>
<td>Compute median early precision. Calls <a class="reference internal" href="BLEval.html#module-BLEval.computeEarlyPrec" title="BLEval.computeEarlyPrec"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.computeEarlyPrec</span></code></a>.</td>
</tr>
<tr class="row-even"><td>-s, –sepr</td>
<td>Analyze median (signed) early precision for activation and inhibitory edges. <a class="reference internal" href="BLEval.html#module-BLEval.computeSignedEPrec" title="BLEval.computeSignedEPrec"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.computeSignedEPrec</span></code></a>.</td>
</tr>
<tr class="row-odd"><td>-m, –motifs</td>
<td>Compute network motifs in the predicted top-k networks. Calls <a class="reference internal" href="BLEval.html#module-BLEval.computeNetMotifs" title="BLEval.computeNetMotifs"><code class="xref py py-mod docutils literal notranslate"><span class="pre">BLEval.computeNetMotifs</span></code></a></td>
</tr>
</tbody>
</table>
<p>For details about the implementation of <a class="reference internal" href="BLEval.html#module-BLEval" title="BLEval"><code class="xref py py-class docutils literal notranslate"><span class="pre">BLEval</span></code></a> , see <a class="reference internal" href="beeline-developer.html#blevalguide"><span class="std std-ref">Adding a new evaluation technique</span></a> .</p>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Aditya Pratapa

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>