

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Defining an input model &mdash; BEELINE  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> BEELINE
          

          
            
            <img src="_static/beeline.svg" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="BEELINE.html">BEELINE</a></li>
<li class="toctree-l1"><a class="reference internal" href="BoolODE.html">BoolODE</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">BEELINE</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Defining an input model</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/boolode-tutorial.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="defining-an-input-model">
<h1>Defining an input model<a class="headerlink" href="#defining-an-input-model" title="Permalink to this headline">¶</a></h1>
<p>Consider a toy gene regulatory network with a series of mutual
inhibition motifs as illustrated below:</p>
<div class="figure align-center" id="id1">
<img alt="_images/boolode-example.png" src="_images/boolode-example.png" />
<p class="caption"><span class="caption-text">If there is an equal probabiliy of taking either the left hand branch
or the right hand branch starting at (g1=high,g2=high), this network will
result in 8 distinct steady states, with one of the eight extreme genes being
active in each steady state.</span></p>
</div>
<p>BoolODE can be used to generate gene expression data of a sample of cells undergoing
a developmental process controlled by a gene regulatory network with a wiring diagram
as shown above.</p>
<p>The above network diagram can be represented as a Boolean model as follows:</p>
<div class="code text highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Gene</span>  <span class="n">Rule</span>
<span class="n">g1</span>    <span class="p">(</span> <span class="n">g1</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g2</span> <span class="p">)</span>
<span class="n">g2</span>    <span class="p">(</span> <span class="n">g2</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g1</span> <span class="p">)</span>
<span class="n">g11</span>   <span class="p">(</span> <span class="n">g1</span> <span class="ow">or</span> <span class="n">g11</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g12</span> <span class="p">)</span>
<span class="n">g12</span>   <span class="p">(</span> <span class="n">g1</span> <span class="ow">or</span> <span class="n">g12</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g11</span> <span class="p">)</span>
<span class="n">g21</span>   <span class="p">(</span> <span class="n">g2</span> <span class="ow">or</span> <span class="n">g21</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g22</span> <span class="p">)</span>
<span class="n">g22</span>   <span class="p">(</span> <span class="n">g2</span> <span class="ow">or</span> <span class="n">g22</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g21</span> <span class="p">)</span>
<span class="n">g111</span>  <span class="p">(</span> <span class="n">g11</span> <span class="ow">or</span> <span class="n">g111</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g112</span> <span class="p">)</span>
<span class="n">g112</span>  <span class="p">(</span> <span class="n">g11</span> <span class="ow">or</span> <span class="n">g112</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g111</span> <span class="p">)</span>
<span class="n">g121</span>  <span class="p">(</span> <span class="n">g12</span> <span class="ow">or</span> <span class="n">g121</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g122</span> <span class="p">)</span>
<span class="n">g122</span>  <span class="p">(</span> <span class="n">g12</span> <span class="ow">or</span> <span class="n">g122</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g121</span> <span class="p">)</span>
<span class="n">g211</span>  <span class="p">(</span> <span class="n">g21</span> <span class="ow">or</span> <span class="n">g211</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g212</span> <span class="p">)</span>
<span class="n">g212</span>  <span class="p">(</span> <span class="n">g21</span> <span class="ow">or</span> <span class="n">g212</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g211</span> <span class="p">)</span>
<span class="n">g221</span>  <span class="p">(</span> <span class="n">g22</span> <span class="ow">or</span> <span class="n">g221</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g222</span> <span class="p">)</span>
<span class="n">g222</span>  <span class="p">(</span> <span class="n">g22</span> <span class="ow">or</span> <span class="n">g222</span> <span class="p">)</span> <span class="ow">and</span> <span class="ow">not</span> <span class="p">(</span> <span class="n">g221</span> <span class="p">)</span>
</pre></div>
</div>
<p>In order to test BoolODE, copy the rules  above into a file, say <code class="docutils literal notranslate"><span class="pre">multistate.txt</span></code>,
in the <code class="docutils literal notranslate"><span class="pre">/data</span></code> folder in the BoolODE folder.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">BoolODE expects a tab separated rule file! Make sure the
input file has tabs between the <code class="docutils literal notranslate"><span class="pre">Gene</span></code> and <code class="docutils literal notranslate"><span class="pre">Rule</span></code> column.</p>
</div>
<p>Notice that the <code class="docutils literal notranslate"><span class="pre">g1&lt;-&gt;g2</span></code> interaction is central to the behavior of
the model.  Let us choose an initial condition such that these genes
have a medium/high value. Biologically, this can be interpreted as the
initial undifferentiated cells have genes <code class="docutils literal notranslate"><span class="pre">g1</span></code> and <code class="docutils literal notranslate"><span class="pre">g2</span></code> ‘ON’, and
all other genes ‘OFF’.  We can specify this initial state by creating a file with the
following lines:</p>
<div class="code text highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Genes</span> <span class="n">Values</span>
<span class="p">[</span><span class="s1">&#39;g1&#39;</span><span class="p">,</span><span class="s1">&#39;g2&#39;</span><span class="p">]</span>   <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
<p>The first column takes a list of genes formatted like a python list of
strings. The second column takes a list of values to assign to the
list of genes, again formatted as a python list of values. <strong>Any gene not
listed in this file will be assigned an initial value of 0 by default.</strong></p>
<p>Let us save the initial condition specification in
<code class="docutils literal notranslate"><span class="pre">/data/multistate_ics.txt</span></code>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">The initial condition file should also be tab
separated. Also, pay attention to the capitalization
of the column names.</p>
</div>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">We recommend starting off with initial condition values
between 0 and 2, since larger values might cause numerical
instability.</p>
</div>
</div>
<div class="section" id="running-boolode">
<h1>Running BoolODE<a class="headerlink" href="#running-boolode" title="Permalink to this headline">¶</a></h1>
<p>A few more decisions before we get our simulated dataset of gene expression
values:</p>
<ol class="arabic simple">
<li>How long do you want to simulate this model? Let’s choose 8 time units</li>
<li>How many cells do you want to simulate? 2000 seems like a good number.</li>
<li>Where would you like to store the simulation output? Say, <code class="docutils literal notranslate"><span class="pre">./test/</span></code>. BoolODE will create any folders that don’t already exist.</li>
<li>Do you want to speed up the simulations by running simulations in parallel? Yes!</li>
</ol>
<p>One last option that we will add is <code class="docutils literal notranslate"><span class="pre">--sample-cells</span></code>, which will instruct BoolODE
to sample one cell from each simulated trajectory.</p>
<p>This is all we need to call BoolODE on this model! The following
directive puts together all the information collected so far</p>
<div class="code bash highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">src</span><span class="o">/</span><span class="n">BoolODE</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">path</span> <span class="n">data</span><span class="o">/</span><span class="n">multistate</span><span class="o">.</span><span class="n">txt</span> \    <span class="c1"># path to the Boolean rules file</span>
                      <span class="o">--</span><span class="n">ics</span> <span class="n">data</span><span class="o">/</span><span class="n">multistate_ics</span><span class="o">.</span><span class="n">txt</span> \ <span class="c1"># path to initial condition specification</span>
                      <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">time</span> <span class="mi">8</span> \                  <span class="c1"># maximum time of simulation</span>
                      <span class="o">--</span><span class="n">num</span><span class="o">-</span><span class="n">cells</span> <span class="mi">2000</span> \              <span class="c1"># number of cells/simulations to perform</span>
                      <span class="o">--</span><span class="n">outPrefix</span> <span class="n">test</span><span class="o">/</span> \             <span class="c1"># relative path to destination</span>
                      <span class="o">--</span><span class="n">do</span><span class="o">-</span><span class="n">parallel</span> \                 <span class="c1"># do parallel simulations</span>
                      <span class="o">--</span><span class="n">sample</span><span class="o">-</span><span class="n">cells</span>                  <span class="c1"># sample one cell from each simulation</span>
</pre></div>
</div>
<p>For a full list of available options, see <a class="reference internal" href="BoolODE.html#boolodeoptions"><span class="std std-ref">Command Line Options</span></a></p>
</div>
<div class="section" id="working-with-boolode-output">
<h1>Working with BoolODE output<a class="headerlink" href="#working-with-boolode-output" title="Permalink to this headline">¶</a></h1>
<p>The simulations will take about 5 minutes to complete. At the end of a successful run, the
output directory should like this:</p>
<div class="code text highlight-default notranslate"><div class="highlight"><pre><span></span>test/                      # User specified destination
|-- parameters.txt         # Parameter names and values generated for input model
|-- ExpressionData.csv     # Gene expression dataset
|-- PseudoTime.csv         # Simulation time of each sample time point/cell
|-- refNetwork.csv         # Boolean network represented as an edge list, the ground truth network
`-- simulations/
    |-- E1.csv
    |-- E2.csv
    ...
</pre></div>
</div>
<p>Where <code class="docutils literal notranslate"><span class="pre">E1.csv,</span> <span class="pre">E2.csv,</span> <span class="pre">...</span></code> are individual simulations. Each column in these
files, the cell IDs, has the form <code class="docutils literal notranslate"><span class="pre">E&lt;simulation</span> <span class="pre">number&gt;_&lt;timepoint&gt;</span></code>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">By default BoolODE will store the entire simulated time
course for every simulation. Specifying the <code class="docutils literal notranslate"><span class="pre">--sample-cells</span></code>
option will result in BoolODE sampling cells and creating an
<code class="docutils literal notranslate"><span class="pre">ExpressionData.csv</span></code> file with genes as rows, and cells as columns.
If the option is not specified, you can write a custom script to
sample cells from each simulation</p>
</div>
<p>In order to visualize the entire dataset, we can carry out dimensionality reduction using
t-SNE. A script like the one below is a good starting point for this.</p>
<div class="code python highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">from</span> <span class="nn">sklearn.manifold</span> <span class="k">import</span> <span class="n">TSNE</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="k">def</span> <span class="nf">vis</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">p</span><span class="p">):</span>
    <span class="n">tsne</span> <span class="o">=</span> <span class="n">TSNE</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span><span class="n">perplexity</span><span class="o">=</span><span class="n">p</span><span class="p">)</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">df</span><span class="o">.</span><span class="n">T</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    <span class="n">tdf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">tsne</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;t-SNE 1&#39;</span><span class="p">,</span> <span class="s1">&#39;t-SNE 2&#39;</span><span class="p">],</span><span class="n">index</span><span class="o">=</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span>
    <span class="n">tdf</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s1">&#39;test/tsne.csv&#39;</span><span class="p">)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">tsne</span><span class="p">[:,</span><span class="mi">0</span><span class="p">],</span> <span class="n">tsne</span><span class="p">[:,</span><span class="mi">1</span><span class="p">],</span> <span class="n">c</span> <span class="o">=</span> <span class="p">[</span><span class="nb">float</span><span class="p">(</span><span class="n">col</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="s1">&#39;test/tree.png&#39;</span><span class="p">)</span>

<span class="n">df</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;test/ExpressionData.csv&#39;</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">vis</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="mi">400</span><span class="p">)</span>
</pre></div>
</div>
<p>Here, the time point of each cell is inferred from the cell ID, and this information
is used to color each cell in the scatter plot. Darker colors imply early time points in the
simulation. The output should look like the following</p>
<div class="figure align-center" id="id2">
<img alt="_images/tree.png" src="_images/tree.png" />
<p class="caption"><span class="caption-text">t-SNE visualization of BoolODE output</span></p>
</div>
<p>Notice the eight steady state clusters! This dataset can now be
processed further using the tools described in  <a class="reference internal" href="BoolODE.html#geninputs"><span class="std std-ref">Generating inputs for BEELINE</span></a> to produce
input datasets for the BEELINE pipeline.</p>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Aditya Pratapa

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>